test/integration/auth/auth.prose.test.ts

```diff
@@ -0,0 +1,416 @@
+              const username = encodeURIComponent(user.username);
+              const password = encodeURIComponent(user.password);
+
+              const url = `${makeConnectionString(
+                this.configuration,
+                username,
+                password
+              )}authMechanism=${mechanism}`;
+
+              const client = this.configuration.newClient(url);
+              try {
+                const stats = await client.db('test').stats();
+                expect(stats).to.exist;
+              } finally {
+                await client.close();
+              }
+            }
+          );
+        }
+
+        it(
+          `authenticates ${user.description} using mechanism negotiaton`,
+          metadata,
+          async function () {
+            const options = {
+              auth: {
+                username: user.username,
+                password: user.password
+              },
+              authSource: 'admin'
+            };
+
+            const client = this.configuration.newClient({}, options);
+            try {
+              const stats = await client.db('test').stats();
+              expect(stats).to.exist;
+            } finally {
+              await client.close();
+            }
+          }
+        );
+
+        it(
+          `authenticates ${user.description} using mechanism negotiaton and url`,
+          metadata,
+          async function () {
+            const username = encodeURIComponent(user.username);
+            const password = encodeURIComponent(user.password);
+            const url = makeConnectionString(this.configuration, username, password);
+
+            const client = this.configuration.newClient(url);
+            try {
+              const stats = await client.db('test').stats();
+              expect(stats).to.exist;
+            } finally {
+              await client.close();
+            }
+          }
+        );
+      }
+
+      /**
+       * Step 2
+       * For a test user supporting both SCRAM-SHA-1 and SCRAM-SHA-256, drivers should verify
+       * that negotation selects SCRAM-SHA-256. This may require monkey patching, manual log
+       * analysis, etc.
+       * todo(NODE-5629): Test passes locally but will fail on CI runs.
+       */
+      it.skip(
+        'selects SCRAM-SHA-256 for a user that supports both auth mechanisms',
+        metadata,
+        async function () {
+          const options = {
+            auth: {
+              username: userMap.both.username,
+              password: userMap.both.password
+            },
+            authSource: this.configuration.db
+          };
+
+          const spy = sinon.spy(ScramSHA256.prototype, 'auth');
+          const client = this.configuration.newClient({}, options);
+          try {
+            const stats = await client.db('test').stats();
+            expect(stats).to.exist;
+            expect(spy.called).to.equal(true);
+          } finally {
+            sinon.restore();
```

Comment:
Can we `sinon.restore` in the afterEach? newClient could throw preventing this cleanup. calling restore if there was no use of sinon is harmless so might as well put it where we can be sure it will run.

Reply:
Moved to the `afterEach` now.